### PR TITLE
[VsCoq1] Progress on #134: Use dune coq top (when enabled by a setting)

### LINKED ISSUE #278: vscoq passes incorrect `-topfile` argument
vscoq uses `-topfile file:///path` or `-topfile untitled:/URL`,  but Coq demands a path (relative or absolute, doesn't matter); that path is compared against the `-Q`/`-R` options to compute the correct name of  the module being created.

For instance, assume `_CoqProject` consists of `-Q theories D`, and `test_topfile.v` is the following:
```coq
Definition hello := Prop.
About hello.
Check D.test_topfile.hello.
```
This code (based on the testcase in https://github.com/ProofGeneral/PG/issues/407#issuecomment-447289254) compiles fine and works in `coqtop` with the correct options, but VSCoq fails to step through it.

Example incorrect command lines that VSCoq uses:
```
exec: /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/bin/coqidetop.opt -main-channel 127.0.0.1:58104:58105 -control-channel 127.0.0.1:58106:58107 -async-proofs on -Q theories D -w -notation-overridden -w -redundant-canonical-projection -w -convert_concl_no_check -w -undeclared-scope -w -ambiguous-paths -w -deprecated-hint-rewrite-without-locality -w -deprecated-instance-without-locality -w -deprecated-typeclasses-transparency-without-locality -topfile untitled:/Users/pgiarrusso1/git/gDOT/dot-iris-make/theories/test_topfile.v
```

```
exec: /Users/pgiarrusso1/.opam/__coq-platform.2021.11.0~8.14~2021.11/bin/coqidetop.opt -main-channel 127.0.0.1:58158:58155 -control-channel 127.0.0.1:58156:58157 -async-proofs on -Q theories D -w -notation-overridden -w -redundant-canonical-projection -w -convert_concl_no_check -w -undeclared-scope -w -ambiguous-paths -w -deprecated-hint-rewrite-without-locality -w -deprecated-instance-without-locality -w -deprecated-typeclasses-transparency-without-locality -topfile file:///Users/pgiarrusso1/git/gDOT/dot-iris-make/theories/test_topfile.v
```

### PR DESCRIPTION
**Pull request by Blaisorblade**
Address #134 for VsCoq1. The first commit fixes #278.

This is a V0 prototype with some limitations, but it appears somewhat suitable for use.
- There's no auto-detection of `dune` projects, but an explicit setting. Since it can be set per workspace, since most dune projects _also_ have a `_CoqProject` file, and since `dune coq top` is probably somewhat experimental, I consider this sufficient for now.
- If the build fails, this does not redirect build outputs/errors to some proper log window (beyond "Coq Language Server" logs).
- User experience isn't great when dune decides to build code: a standard progress bar appears, but one must view "Coq Language Server" to understand what is happening.

What I tested was my integration branch — https://github.com/Blaisorblade/vscoq/tree/thread-count-option-misc.

Some unexpected messages appeared, but they seem due to the options added by dune coq top.
```
coqtop-stderr: Warning: Native compiler is disabled, -native-compiler ondemand option
ignored. [native-compiler-disabled,native-compiler]
Warning: The native-compiler option is deprecated. To compile native files
ahead of time, use the coqnative binary instead.
[deprecated-native-compiler-option,deprecated]

coqtop-stdout:Skipping rcfile loading.
```

The last one suggests that `.coqrc` is not being loaded, but it appears untrue — I used
```
Test Nested Proofs Allowed.
Test Ltac Backtrace.
```
to confirm my  `.coqrc` settings were active.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -56,10 +56,15 @@
...
+        "coqtop.dunePath": {
+          "type": "string",
+          "default": "dune",
+          "markdownDescription": "Full path to dune binary (`dune` included)."
+        },
         "coqtop.binPath": {
           "type": "string",
           "default": "",
-          "description": "Path to coqtop and coqidetop binaries (alternatively, hoqtop and hoqidetop)."
+          "description": "Path to folder with coqtop and coqidetop binaries (alternatively, hoqtop and hoqidetop)."
```
Comment (Blaisorblade) [inline package.json:67]: This is an unrelated message improvement.

File: package.json [hunk 2]
```diff
@@ -85,6 +90,10 @@
           "default": "open-script",
           "markdownDescription": "When to start an instance of coqtop: when a Coq script is opened (`open-script`) or else when the user begins interaction (`interaction`; default)."
         },
+        "coqtop.useDune": {
+          "type": "boolean",
+          "markdownDescription": "Whether to use `dune coqtop`"
```
Comment (Blaisorblade) [inline package.json:95]: This description must be expanded, and the feature marked experimental.

**Comment by Blaisorblade** [comment]
This isn't complete but I've been using it for a while (must still rebase over 0.3.7).

I'd like to ask what would be considered necessary. The lack of a decent progress dialog seems the most annoying problem...

**Comment by Blaisorblade** [comment]
Bug report: this still passes flags from `_CoqProject` to Coq. That's one I can fix.

**Comment by 4ever2** [comment]
Doesn't this PR also fix #278?

**Comment by Blaisorblade** [comment]
Yep, thanks @4ever2

File: package.json (hunk 2 above)
Comment (Blaisorblade) [reply package.json]: Marked experimental. (Not sure what expansion I had in mind).

**Comment by thery** [comment]
@Blaisorblade tell me when you think it is mergeable

**Comment by huynhtrankhanh** [comment]
This PR is not ready to merge I guess. But one nitpick: why should we continue using `var`? We should use `const` and `let`, and prefer `const` whenever possible. `var` is function scoped, not block scoped like the other two keywords. This is very inconsequential for this PR though.

I'll probably review this PR more carefully later.

File: server/src/coqtop/CoqTop8.ts [hunk 3]
```diff
@@ -227,16 +239,27 @@ export class CoqTop extends IdeSlave8 implements coqtop.CoqTop {
...
         '-async-proofs', 'on'
         ].concat(this.settings.args);
     }
-    this.console.log('exec: ' + coqtopModule + ' ' + args.join(' '));
-    return spawn(coqtopModule, args, {detached: false, cwd: this.projectRoot});
+    if (this.settings.useDune) {
+      var binary = this.settings.dunePath
+      if (scriptPath === undefined)
+        throw new CoqtopSpawnError("", "File was not saved to local file system");
+      var args = ["coq", "top", "--toplevel=" + coqtopModule,
```
Comment (huynhtrankhanh) [inline server/src/coqtop/CoqTop8.ts:254]: Use `const`. Also use ES6 spread syntax instead of `concat`.

**Review by huynhtrankhanh — CHANGES_REQUESTED** [review]

File: server/src/coqtop/CoqTop8.ts [hunk 4]
```diff
@@ -210,15 +211,26 @@ export class CoqTop extends IdeSlave8 implements coqtop.CoqTop {
     return path.join(this.settings.binPath.trim(), this.settings.coqidetopExe);
   }
 
+  private get scriptPath() {
+    // TODO: rename scriptFile to scriptUri everywhere.
+    let uri = Uri.parse(this.scriptFile)
+    if (uri.scheme == "file")
+      return uri.fsPath;
+    else
+      return undefined
```
Comment (Blaisorblade) [inline server/src/coqtop/CoqTop8.ts:220]: This is new since last revision and enables stepping through unsaved files (but only outside dune coq top).

File: server/src/coqtop/CoqTop8.ts [hunk 5]
```diff
@@ -250,17 +250,16 @@ export class CoqTop extends IdeSlave8 implements coqtop.CoqTop {
         ...this.settings.args
         ];
```
Comment (Blaisorblade) [inline server/src/coqtop/CoqTop8.ts:251]: I haven't ported this code to `const`, but it should be doable the same way. (But I prefer to avoid mixing refactorings with changes....).

File: server/src/coqtop/CoqTop8.ts [hunk 6]
```diff
@@ -250,17 +250,16 @@ export class CoqTop extends IdeSlave8 implements coqtop.CoqTop {
...
+
+    if (this.settings.useDune && scriptPath === undefined) {
+      throw new CoqtopSpawnError("", "File was not saved to local file system");
     }
 
+    const [binary, args] = this.settings.useDune ?
+      [this.settings.dunePath, ["coq", "top", "--toplevel=" + coqtopModule,
+        scriptPath, "--", ...coqArgs]] :
+      [coqtopModule, coqArgs]
+
```
Comment (Blaisorblade) [inline server/src/coqtop/CoqTop8.ts:262]: @huynhtrankhanh Since the code relied on `var`'s scoping, this seems the smallest possible change.

Had to lift error checking since TS seems to lack statement expressions, but I guess that's not too bad. I guess `function () { ... } ()` lets you encode them, but I wouldn't know if that's idiomatic.

Comment (Blaisorblade) [reply server/src/coqtop/CoqTop8.ts:262]: To be clear, if anything seems silly it's because I'm not really a TypeScript dev and I was following existing idioms.

**Comment by Blaisorblade** [comment]
I've done the main changes I planned, tried to address @huynhtrankhanh's comments, but I'll be retesting more later today.

File: server/src/coqtop/CoqTop8.ts (hunk 5 above)
Comment (Blaisorblade) [reply server/src/coqtop/CoqTop8.ts]: Done in last commit (untested yet).

**Comment by Blaisorblade** [comment]
The latest changes seem to work, and I addressed @huynhtrankhanh 's comments, so I consider this ready for review and merging.

**Comment by thery** [comment]
@Blaisorblade Perfect I will try to have a look at it today. Would be good if @huynhtrankhanh would give his opinion too

File: server/src/CoqProject.ts [hunk 7]
```diff
@@ -86,9 +86,12 @@ export class CoqProject {
       this.coqProjectRoot = newSettings.coq.coqProjectRoot;
       this.console.log("Updated project root to " + this.getCoqProjectRoot());
     }
-    if(newSettings.coq.loadCoqProject ) {
+    if(newSettings.coq.loadCoqProject && !newSettings.coqtop.useDune) {
```
Comment (thery) [inline server/src/CoqProject.ts:89]: does this means that if your flag is not set, everything works like as before?

**Review by thery — APPROVED** [review]

**Comment by thery** [comment]
@Blaisorblade  I had a look, it seems ok.
As soon as @huynhtrankhanh is ok, we can merge

**Comment by huynhtrankhanh** [comment]
This PR looks perfect. I really appreciate your work. However, there's one issue: the configuration options! There are two options here, which should be mutually exclusive: `coqtop.loadCoqProject` and `coqtop.useDune`. I believe having two mutually exclusive options that could be set independently causes problems when we further build open your work and makes it hard for users to predict the effect of their configuration. Should we have a single option that allows us to choose whether to load _CoqProject, use Dune or ignore both?

**Comment by thery** [comment]
@huynhtrankhanh you would prefer a coqtop.configMode  with the options coqproject or dune?

File: server/src/CoqProject.ts (hunk 7 above)
Comment (Blaisorblade) [reply server/src/CoqProject.ts:89]: Yes, that was my design goal :-).

**Comment by Blaisorblade** [comment]
Thanks for the review!

> Should we have a single option that allows us to choose whether to load _CoqProject, use Dune or ignore both?

Good idea; that works for me. I'm afraid I can get back to it in a couple days.

Dropping "ignore both" as @thery suggests also seems plausible but has a slightly higher chances of regressions — does any user _need_ to disable `loadCoqProject`? We should at least test projects with _no_ `_CoqProject` file.

**Comment by thery** [comment]
@Blaisorblade @huynhtrankhanh I would like to make a new release beginning of march, do you think we can make this PR ready to merge by then?

**Comment by Blaisorblade** [comment]
Sorry I'm late — I plan to address the comment next week, what I pushed just updates the extension for dune coq top changes (now dune wants relative paths, so we oblige).

**Comment by thery** [comment]
what is the status of this PR?

**Comment by Blaisorblade** [comment]
Sorry this hasn't progressed :-| ... Closing for now.
